pkg/handler/enqueue_mapped.go

```diff
@@ -73,7 +73,11 @@ func (e *enqueueRequestsFromMapFunc) Generic(evt event.GenericEvent, q workqueue
 }
 
 func (e *enqueueRequestsFromMapFunc) mapAndEnqueue(q workqueue.RateLimitingInterface, object client.Object) {
+	reqs := map[reconcile.Request]bool{}
```

Comment:
I don't think a simple map is enough here, what we need is **an ordered set**, just like the `RequestSet` implementation in #1380 .

Reply:
I think that's a question of API promises. The return value from the map function is an array but do we consider that a promise that things will end up in the workqueue in that order? It's non-deterministic how things will actually run across multiple workers, though overall ordering would be respected. I wouldn't think that reconciles within the same controller (i.e. of the same root object type) would be order sensitive in general though.

Reply:
Got your point, but if the user-defined map function will generate an order-sensitive array, should we shuffle the order with a map even for the deduplication?

Reply:
I defer to the wisdom of the group :) I wouldn't consider order to be preserved there since it's not really in line with the convergence-based aesthetic of controllers. But we can totally do it if we want to spend the CPU cycles on slightly more tracking :D

Reply:
So first off, there is no obvious rule if if we were to sort this the result of oldObj or newObj comes first. Second, deduplicating is already an microoptimization that should be irrelevant for most controllers (if it isn't, you should probably reconsider the architecture of your controller, you shouldn't be blocking for long-running things in `Reconcile`). So I am opposed to make ordering-when-deduplicating part of the api and adding additional code for that.

Reply:
I did actually change it to preserve the ordering on mapfuncs because it was easy :)